### PR TITLE
Implement Hyperparameter tuning for Gaussian Process surrogate modeling

### PR DESCRIPTION
**Pull request by iwhalvic**
<!--
This implement hyperparameter tuning into the Gaussian Process surrogate model. This should vastly expand their usefulness as is removes a significant portion of the user burden of model selection. Tuning is implemented for 3 covariance function, which each covariance function cable of tuning the noise_variance, signal_variance, and a vector of length_factors.

length_factor tuning is of particular interest, as it can be used to perform some basic Automatic Relevance Determination (ARD)

Tuning of specific hyperparameters can be toggled (such as disabling noise_variance tuning). Similarly tuning bounds can also be user specified.

The driver of this tuning is an optimization of an objective function in the hyparameter space. To do this we leverage the conjugate-gradient solvers in the PETSc/TAO package. The selected objective function is the negative log-likelihood.

### CODE REVIEW

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md:64]: These figures are pretty big. You can make them smaller by doing `style=width:60%`. But I will leave this to your discretion.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md:98]: It would be good to mention what input file is running this example. You don't need to show a `!listing` if you don't want.

**Review by zachmprince — CHANGES_REQUESTED** [review]
This looks like it's almost there. Great job! A couple things on the example inputs. Make sure you are only committing files and producing results that are used in the example. I noticed that there are few more files than what are shown and some of them are creating a surrogate for the max temperature. Second, all input files need to be tested, so you need to have a tests file in your example folder. If the examples take a long time to run, you can reduce the size using `cli_args` or just doing `check_input = true`, which will basically just test the input syntax.

File: modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C:65]: I included the error exist because if a user types into the tunabel list wrong sigmal_variance instead of signal_variance, they get en error directing them to look, rather than just not throwing anything. This little framework specifies the parameters you wish to tune as a list, so it bypasses some of the standard parameter error checking.

File: modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C:78]: I believe I just need a "large" number, I just changed it to 1e9. Users can override this in the parameters if needed anyways.

PETSC_INFINITY is just a default used by TAO, so I tried to keep consistency with that. But not a big deal.

File: modules/stochastic_tools/include/surrogates/CovarianceFunctionBase.h
Comment (iwhalvic) [reply modules/stochastic_tools/include/surrogates/CovarianceFunctionBase.h:43]: I added a small function for this in the base class. Its basically just to spit out an error if the user is attempting to perform hyperparameter tuning on a covariance function that does not have appropriate derivatives defined. To me it seems conceivable a user may want to set up a covariance function to use, but may not wish to use hyperparameter tuning, and thus would not need to define this function in their overridden class.

File: modules/stochastic_tools/src/surrogates/ExponentialCovariance.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/ExponentialCovariance.C:33]: See the earlier comment. What hyperparams are tunable is a matter of the defined derivatives, which is based on the selection of covariance function. So to me it seems better to have these flags set in the derived class.

Ill admit I have sort of went back on forth on where these three parameters should live, as they are extremely common to covariance functions (but not strictly necessary). At the moment I have left the actual user input of the variables in the base class, but anything dealing with how they are used  is in the derived classes.

File: modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C:31]: Ok. 

noise_variance = 0 as a default, looking at data is a noiseless way makes sense in certain scenarios.

Other two I made required. No obvious way to me at the moment to define them.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:44]: Good call. I had that in during an experimental stage before I had tune_parameter working for parm-by-parm tuning.

File: modules/stochastic_tools/src/surrogates/MaternHalfIntCovariance.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/MaternHalfIntCovariance.C:42]: added `addAdditionalHyperParamMap` and ``loadAdditionalHyperParamMap``

File: modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C:19]: Added error checks in `computeCovarianceMatrix`

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:96]: See other response. At the moment I still have the error in the original spot. Basically there can be two reasons why a parameter is not tunable. It may be a valid parameter in the list, however tuning is not enables for that parameter.

However, it could just be a nonsense/mistyped name.

Regardless, it should return False.

I can rework this and pull the isParamValid out of isTunable, but I am not sure if that is any better.

File: modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C
Comment (zachmprince) [reply modules/stochastic_tools/src/surrogates/CovarianceFunctionBase.C]: So you can use `std::numeric_limits<Real>::max()`. This makes things less arbitrary.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (zachmprince) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:95]: I agree with you now. I was not considering the fact that a parameter might be valid, but not tunable.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C]: I have made some changes, "testing" by swapping `ifdef `with `ifndef`, let me know if we need to put even more safeguards in place.

File: modules/stochastic_tools/src/vectorpostprocessors/GaussianProcessData.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/vectorpostprocessors/GaussianProcessData.C:13]: There is a use case for users to view the hyperparameter data, particularly the length factor.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md
Comment (iwhalvic) [reply modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md:2]: I added a small introductory blurb. Not sure if this is exactly what you are wanting.

**Comment by zachmprince** [comment]
@iwhalvic could you do a rebase and a clang format so we can to the testing?

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md:71]: Still need to address this.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/gaussian_process_surrogate.md:59]:
```suggestion
One advantage of Gaussian Process surrogates is the ability to provide an model for uncertainty. To output this data the standard `EvaluateSurrogate` vector post processor is replaced with the [EvaluateGaussianProcess](EvaluateGaussianProcess.md) vector post processor, which functions similarly but also outputs the standard deviation of the surrogate at the evaluation point.  
```

File: modules/stochastic_tools/examples/surrogates/gaussian_process/GP_normal_mc.i
Comment (zachmprince) [inline modules/stochastic_tools/examples/surrogates/gaussian_process/GP_normal_mc.i:113]: Remove the commented lines and update the description.

File: modules/stochastic_tools/examples/surrogates/gaussian_process/GP_normal_mc.i
Comment (zachmprince) [inline modules/stochastic_tools/examples/surrogates/gaussian_process/GP_normal_mc.i:121]: Same here

File: modules/stochastic_tools/examples/surrogates/gaussian_process/sub.i
Comment (zachmprince) [inline modules/stochastic_tools/examples/surrogates/gaussian_process/sub.i:66]: Need to add tests for all the inputs in the examples folder. MOOSE SQA says we can't check in input files without adding tests.

File: modules/stochastic_tools/include/surrogates/CovarianceFunctionBase.h
Comment (zachmprince) [inline modules/stochastic_tools/include/surrogates/CovarianceFunctionBase.h:53]: Don't comment out the parameter names if they are being used in the definition.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:230]:
```suggestion
  ierr = TaoCreate(_tao_comm, &tao);
```

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:223]: What is the purpose of the return value here?

File: modules/stochastic_tools/test/tests/surrogates/gaussian_process/tests
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/surrogates/gaussian_process/tests:39]:
```suggestion
    detail = 'and training and evaluating with a Matern half integer kernel, '
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Mostly small things. One big thing is you need to add a tests file for your inputs in the example.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C
Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:223]: This was due to PETSc CHKERRQ behavior, I have switched to CHKERRV, which i think is appropriate in this situation. See here https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/CHKERRQ.html#CHKERRQ

Comment (zachmprince) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:223]: I think we should self contain all the petsc stuff to this function. So keeping this function as void and having a `CHKERRQ` within in. What do you think?

Comment (iwhalvic) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:223]: Update: Retained CHKERRQ and added an error check on hyperparamTuning() return

File: modules/stochastic_tools/examples/surrogates/gaussian_process/sub.i
Comment (iwhalvic) [reply modules/stochastic_tools/examples/surrogates/gaussian_process/sub.i:66]: Added tests. Attempted to model off example tests already in there (poly chaos). Marked full model test as heavy.

**Comment by aeslaughter** [comment]
I am unable to reproduce the documentation error. I am running the target again to see what happens.

**Comment by aeslaughter** [comment]
I am going to wait for the next branch to be fixed up before this is merged.

**Review by zachmprince — APPROVED** [review]

**Comment by loganharbour** [comment]
Someone is going to need to work on this: https://civet.inl.gov/job/626935/

Let me know if you need help.

**Pull request merged**
